subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CodeNarcPlugin.java

```diff
@@ -80,59 +75,32 @@ protected void configureTaskDefaults(CodeNarc task, String baseName) {
     }
 
     private void configureDefaultDependencies(Configuration configuration) {
-        configuration.defaultDependencies(new Action<DependencySet>() {
-            @Override
-            public void execute(DependencySet dependencies) {
-                dependencies.add(project.getDependencies().create("org.codenarc:CodeNarc:" + extension.getToolVersion()));
+        configuration.defaultDependencies(dependencies -> {
+            dependencies.add(project.getDependencies().create("org.codenarc:CodeNarc:" + extension.getToolVersion()));
+            if (extension.getToolVersion().equals("1.5")) {
```

Comment:
I did not use `DEFAULT_CODENARC_VERSION` constant here because I see this as a workaround and once a new version of CodeNarc tool comes out, we will update the constant and should check if we need to force something or whether it will work out of the box.

Reply:
This isn't quite what I had in mind.  I was thinking that in our test coverage, we could force Groovy to 2.5.10 when running on JDK14, but leave the plugin unopinionated otherwise and use whatever version of Groovy CodeNarc asks for on anything <JDK14.  CodeNarc 1.5 is only on Groovy 2.3, so an upgrade to 2.5.10 might be a long time coming.

If we want to make CodeNarc work out-of-the-box with 6.3 and JDK14, then I think we should check if we're running on JDK14 here and then add a constraint to require at least Groovy 2.5.10.  I don't think we have to limit this to just CodeNarc 1.5 if other versions may work.  I think we also need to fix the versions of other dependencies because CodeNarc also has direct dependencies on groovy-xml and groovy-ant, which I think need to align with the Groovy version. 

@jjohannes do you have a suggestion on how to do this with the new rich version constraints?  Basically, we would prefer to use what CodeNarc wants transitively, unless we're on JDK14, then we'd like to use at least 2.5.10.

Reply:
I think I don't have all the context. So I am not sure: are we talking about adding something to production code or test code?

As I read it, you don't want to "force" a versions (as the code below suggests). You just want a minimal version under a certain condition. So you should add a 'normal' version constraint (i.e. a constraint to a _required_ version) for that minimal version. 

```
if (jdkVeryion >= 14) {
  project.getDependencies().getConstraints().add(configuration.getName(), "org.codehaus.groovy:groovy:2.5.10")
}
```

Reply:
Added constraint as suggested.

In case we only want to workaround this for JDK14 in tests by forcing a later groovy version: https://github.com/gradle/gradle/pull/12420